File: backend/app/providers/ai/mock.py
"""Development AI provider.

This is NOT a trained model and makes no claim to be one. It derives a stable,
plausible-looking detection set from cheap image statistics (dark-region ratio,
blue dominance, contrast) seeded by a hash of the image bytes, so that:

* the same photo always yields the same analysis - demos and tests are stable;
* visibly different photos yield visibly different analyses;
* the shape of the output exactly matches what the real detector returns.

Swap it for the real thing by setting ``AI_PROVIDER=http`` and pointing
``AI_SERVICE_URL`` at the ai-service container. See ``docs/AI.md``.
"""

from __future__ import annotations

import hashlib
import io
import random
import time

from PIL import Image, ImageFilter, UnidentifiedImageError

from app.core.enums import DamageType
from app.providers.ai.base import AnalysisResult, Detection

#: Damage classes the development stub can emit, with the relative likelihood
#: used when image statistics do not clearly point at one class.
_BASE_WEIGHTS: list[tuple[DamageType, float]] = [
    (DamageType.POTHOLE, 0.40),
    (DamageType.CRACKED_ROAD, 0.25),
    (DamageType.FLOODING, 0.12),
    (DamageType.DAMAGED_SIDEWALK, 0.13),
    (DamageType.BROKEN_STREETLIGHT, 0.10),
]


class MockAIProvider:
    """Deterministic stand-in for the vision model."""

    name = "mock"

    def __init__(self, model_name: str = "roadwatch-dev-stub", model_version: str = "0.1.0") -> None:
        self.model_name = model_name
        self.model_version = model_version

    async def analyze(self, image_bytes: bytes, content_type: str) -> AnalysisResult:
        started = time.perf_counter()
        try:
            stats = _image_stats(image_bytes)
        except (UnidentifiedImageError, OSError, ValueError):
            return AnalysisResult(
                damage_type=DamageType.UNKNOWN,
                confidence=0.0,
                provider=self.name,
                model_name=self.model_name,
                model_version=self.model_version,
                processing_ms=int((time.perf_counter() - started) * 1000),
                error_message="The uploaded file could not be read as an image.",
            )

        seed = int(hashlib.sha256(image_bytes).hexdigest()[:16], 16)
        rng = random.Random(seed)

        primary = _choose_damage_type(stats, rng)
        detections = _build_detections(primary, stats, rng)

        if not detections:
            return AnalysisResult(
                damage_type=DamageType.UNKNOWN,
                confidence=0.0,
                provider=self.name,
                model_name=self.model_name,
                model_version=self.model_version,
                processing_ms=int((time.perf_counter() - started) * 1000),
            )

        best = max(detections, key=lambda d: d.confidence)
        # Overlapping boxes must not push coverage above the frame.
        area_ratio = min(1.0, sum(d.area_ratio for d in detections))

        return AnalysisResult(
            damage_type=best.damage_type,
            confidence=best.confidence,
            detections=detections,
            damaged_area_ratio=round(area_ratio, 4),
            provider=self.name,
            model_name=self.model_name,
            model_version=self.model_version,
            processing_ms=int((time.perf_counter() - started) * 1000),
        )

    async def health(self) -> bool:
        return True


class _Stats:
    __slots__ = ("dark_ratio", "blue_dominance", "edge_density", "brightness", "top_light_ratio")

    def __init__(
        self,
        dark_ratio: float,
        blue_dominance: float,
        edge_density: float,
        brightness: float,
        top_light_ratio: float,
    ):
        self.dark_ratio = dark_ratio
        self.blue_dominance = blue_dominance
        self.edge_density = edge_density
        self.brightness = brightness
        self.top_light_ratio = top_light_ratio


def _image_stats(image_bytes: bytes) -> _Stats:
    """Cheap 64x64 summary of the photo used to bias the stub's output."""
    with Image.open(io.BytesIO(image_bytes)) as img:
        img.load()
        rgb = img.convert("RGB").resize((64, 64))

    pixels = list(rgb.getdata())
    count = len(pixels)
    luminance = [(0.299 * r + 0.587 * g + 0.114 * b) for r, g, b in pixels]
    brightness = sum(luminance) / count

    # A pothole reads as a dark patch against lighter tarmac.
    threshold = max(20.0, brightness * 0.62)
    dark_ratio = sum(1 for lum in luminance if lum < threshold) / count

    blue_dominance = sum(1 for r, g, b in pixels if b > r + 12 and b > g + 6) / count

    grey = rgb.convert("L").filter(ImageFilter.FIND_EDGES)
    edge_pixels = list(grey.getdata())
    edge_density = sum(1 for value in edge_pixels if value > 40) / len(edge_pixels)

    # Paved footpath reads as a broad, bright band across the top of the frame.
    top_rows = luminance[: count // 2]
    top_light_ratio = sum(1 for lum in top_rows if lum > 125) / max(1, len(top_rows))

    return _Stats(dark_ratio, blue_dominance, edge_density, brightness / 255.0, top_light_ratio)


def _choose_damage_type(stats: _Stats, rng: random.Random) -> DamageType:
    """Pick a class from visual evidence, falling back to chance only when the
    evidence is genuinely ambiguous.

    The thresholds are crude by design - this is a stub, not a detector - but
    the ordering reflects how separable each cue is: standing water and a
    night-time scene are unmistakable, a dark patch on tarmac is a strong
    pothole cue, and dense fine edges without a dark patch suggest cracking.
    """
    if stats.blue_dominance > 0.20:
        return DamageType.FLOODING
    if stats.brightness < 0.25:
        return DamageType.BROKEN_STREETLIGHT
    if stats.top_light_ratio > 0.40:
        return DamageType.DAMAGED_SIDEWALK

    # Cracking vs a pothole is the one genuinely close call. A pothole is a
    # compact dark blob: lots of dark pixels relative to edge content. Cracking
    # is the opposite - dense fine edges with little dark area.
    edge_to_dark = stats.dark_ratio / max(stats.edge_density, 1e-6)
    if stats.edge_density >= 0.125 and edge_to_dark < 0.5:
        return DamageType.CRACKED_ROAD
    if stats.dark_ratio > 0.005:
        return DamageType.POTHOLE
    if stats.edge_density > 0.085:
        return DamageType.CRACKED_ROAD

    # Nothing decisive in the frame - fall back to the prior.
    population = [item[0] for item in _BASE_WEIGHTS]
    return rng.choices(population, weights=[item[1] for item in _BASE_WEIGHTS], k=1)[0]


def _build_detections(primary: DamageType, stats: _Stats, rng: random.Random) -> list[Detection]:
    """Emit 1-3 boxes: the primary class plus an occasional secondary finding."""
    # Standing water covers ground without being dark, so blue coverage has to
    # count towards extent or every flood would look equally minor.
    severity_hint = min(
        1.0, stats.dark_ratio * 1.8 + stats.edge_density * 0.6 + stats.blue_dominance * 0.75
    )
    box_count = 1 + int(severity_hint > 0.45) + int(severity_hint > 0.75)

    detections: list[Detection] = []
    for index in range(box_count):
        # Later boxes are smaller and less certain, as with a real detector.
        scale = 1.0 - (index * 0.28)
        width = round(min(0.62, max(0.06, (0.16 + severity_hint * 0.34) * scale)), 4)
        height = round(min(0.55, max(0.05, width * rng.uniform(0.55, 1.05))), 4)
        detections.append(
            Detection(
                damage_type=primary,
                confidence=round(min(0.98, max(0.30, rng.uniform(0.62, 0.95) * scale + 0.08)), 4),
                bbox_x=round(rng.uniform(0.04, max(0.05, 0.94 - width)), 4),
                bbox_y=round(rng.uniform(0.10, max(0.11, 0.90 - height)), 4),
                bbox_width=width,
                bbox_height=height,
            )
        )

    # Real scenes are often mixed - a flooded stretch also shows cracking.
    if stats.edge_density > 0.34 and primary != DamageType.CRACKED_ROAD and rng.random() < 0.45:
        detections.append(
            Detection(
                damage_type=DamageType.CRACKED_ROAD,
                confidence=round(rng.uniform(0.48, 0.72), 4),
                bbox_x=round(rng.uniform(0.05, 0.55), 4),
                bbox_y=round(rng.uniform(0.15, 0.65), 4),
                bbox_width=round(rng.uniform(0.10, 0.26), 4),
                bbox_height=round(rng.uniform(0.06, 0.18), 4),
            )
        )

    return detections

File: supabase/functions/api/providers/ai/mock.ts
// Ported from backend/app/providers/ai/mock.py. NOT a trained model - derives
// a stable, plausible-looking detection set from cheap image statistics
// (dark-region ratio, blue dominance, edge density) seeded by the image
// bytes, so the same photo always yields the same analysis.
//
// The seeded-RNG algorithm itself (mulberry32, see _shared/random.ts) does
// NOT match Python's Mersenne Twister bit-for-bit - only the qualitative
// contract matters here (deterministic per image, distinguishable outputs),
// since this whole provider is an explicitly-labelled dev stub, never
// production inference.

import type { AIAnalysisProvider, AnalysisResult, Detection } from "./base.ts";
import type { DamageType } from "../../_shared/enums.ts";
import resizeImg from "npm:@jsquash/resize@2.1.1/index.js";
import { decodeAny } from "../../_shared/image_codec.ts";
import { mulberry32, seedFromBytes, uniform, weightedChoice } from "../../_shared/random.ts";

const BASE_WEIGHTS: [DamageType, number][] = [
  ["POTHOLE", 0.40],
  ["CRACKED_ROAD", 0.25],
  ["FLOODING", 0.12],
  ["DAMAGED_SIDEWALK", 0.13],
  ["BROKEN_STREETLIGHT", 0.10],
];

interface Stats {
  darkRatio: number;
  blueDominance: number;
  edgeDensity: number;
  brightness: number;
  topLightRatio: number;
}

function luminanceOf(r: number, g: number, b: number): number {
  return 0.299 * r + 0.587 * g + 0.114 * b;
}

async function imageStats(image: ImageData): Promise<Stats> {
  const small = await resizeImg(image, { width: 64, height: 64, method: "lanczos3", fitMethod: "stretch" });
  const { width, height, data } = small;
  const count = width * height;

  const luminance = new Float64Array(count);
  let sum = 0;
  for (let i = 0; i < count; i++) {
    const o = i * 4;
    const l = luminanceOf(data[o], data[o + 1], data[o + 2]);
    luminance[i] = l;
    sum += l;
  }
  const brightness = sum / count;

  const threshold = Math.max(20.0, brightness * 0.62);
  let darkCount = 0;
  let blueCount = 0;
  for (let i = 0; i < count; i++) {
    if (luminance[i] < threshold) darkCount++;
    const o = i * 4;
    if (data[o + 2] > data[o] + 12 && data[o + 2] > data[o + 1] + 6) blueCount++;
  }

  // 3x3 FIND_EDGES-equivalent convolution over the greyscale image
  // (kernel [-1,-1,-1, -1,8,-1, -1,-1,-1]); border pixels pass through
  // unfiltered, matching PIL's built-in 3x3 filter behaviour.
  let edgeCount = 0;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let edgeValue: number;
      if (x === 0 || y === 0 || x === width - 1 || y === height - 1) {
        edgeValue = 0;
      } else {
        let acc = 8 * luminance[y * width + x];
        for (const [dx, dy] of [[-1, -1], [0, -1], [1, -1], [-1, 0], [1, 0], [-1, 1], [0, 1], [1, 1]]) {
          acc -= luminance[(y + dy) * width + (x + dx)];
        }
        edgeValue = Math.max(0, Math.min(255, acc));
      }
      if (edgeValue > 40) edgeCount++;
    }
  }
  const edgeDensity = edgeCount / count;

  const topRowCount = Math.floor(count / 2);
  let topLightCount = 0;
  for (let i = 0; i < topRowCount; i++) {
    if (luminance[i] > 125) topLightCount++;
  }
  const topLightRatio = topLightCount / Math.max(1, topRowCount);

  return { darkRatio: darkCount / count, blueDominance: blueCount / count, edgeDensity, brightness: brightness / 255, topLightRatio };
}

function chooseDamageType(stats: Stats, rng: () => number): DamageType {
  if (stats.blueDominance > 0.20) return "FLOODING";
  if (stats.brightness < 0.25) return "BROKEN_STREETLIGHT";
  if (stats.topLightRatio > 0.40) return "DAMAGED_SIDEWALK";

  const edgeToDark = stats.darkRatio / Math.max(stats.edgeDensity, 1e-6);
  if (stats.edgeDensity >= 0.125 && edgeToDark < 0.5) return "CRACKED_ROAD";
  if (stats.darkRatio > 0.005) return "POTHOLE";
  if (stats.edgeDensity > 0.085) return "CRACKED_ROAD";

  return weightedChoice(rng, BASE_WEIGHTS.map((w) => w[0]), BASE_WEIGHTS.map((w) => w[1]));
}

function round(value: number, decimals: number): number {
  const f = 10 ** decimals;
  return Math.round(value * f) / f;
}

function buildDetections(primary: DamageType, stats: Stats, rng: () => number): Detection[] {
  const severityHint = Math.min(1.0, stats.darkRatio * 1.8 + stats.edgeDensity * 0.6 + stats.blueDominance * 0.75);
  const boxCount = 1 + (severityHint > 0.45 ? 1 : 0) + (severityHint > 0.75 ? 1 : 0);

  const detections: Detection[] = [];
  for (let index = 0; index < boxCount; index++) {
    const scale = 1.0 - index * 0.28;
    const width = round(Math.min(0.62, Math.max(0.06, (0.16 + severityHint * 0.34) * scale)), 4);
    const height = round(Math.min(0.55, Math.max(0.05, width * uniform(rng, 0.55, 1.05))), 4);
    detections.push({
      damageType: primary,
      confidence: round(Math.min(0.98, Math.max(0.30, uniform(rng, 0.62, 0.95) * scale + 0.08)), 4),
      bboxX: round(uniform(rng, 0.04, Math.max(0.05, 0.94 - width)), 4),
      bboxY: round(uniform(rng, 0.10, Math.max(0.11, 0.90 - height)), 4),
      bboxWidth: width,
      bboxHeight: height,
    });
  }

  if (stats.edgeDensity > 0.34 && primary !== "CRACKED_ROAD" && rng() < 0.45) {
    detections.push({
      damageType: "CRACKED_ROAD",
      confidence: round(uniform(rng, 0.48, 0.72), 4),
      bboxX: round(uniform(rng, 0.05, 0.55), 4),
      bboxY: round(uniform(rng, 0.15, 0.65), 4),
      bboxWidth: round(uniform(rng, 0.10, 0.26), 4),
      bboxHeight: round(uniform(rng, 0.06, 0.18), 4),
    });
  }

  return detections;
}

export class MockAIProvider implements AIAnalysisProvider {
  name = "mock";
  constructor(private modelName = "roadwatch-dev-stub", private modelVersion = "0.1.0") {}

  async analyze(imageBytes: Uint8Array, _contentType: string): Promise<AnalysisResult> {
    const started = performance.now();
    const decoded = await decodeAny(imageBytes);
    if (!decoded) {
      return {
        damageType: "UNKNOWN",
        confidence: 0.0,
        detections: [],
        damagedAreaRatio: 0.0,
        provider: this.name,
        modelName: this.modelName,
        modelVersion: this.modelVersion,
        processingMs: Math.round(performance.now() - started),
        errorMessage: "The uploaded file could not be read as an image.",
      };
    }

    const stats = await imageStats(decoded.image);
    const seed = await seedFromBytes(imageBytes);
    const rng = mulberry32(seed);

    const primary = chooseDamageType(stats, rng);
    const detections = buildDetections(primary, stats, rng);

    if (detections.length === 0) {
      return {
        damageType: "UNKNOWN",
        confidence: 0.0,
        detections: [],
        damagedAreaRatio: 0.0,
        provider: this.name,
        modelName: this.modelName,
        modelVersion: this.modelVersion,
        processingMs: Math.round(performance.now() - started),
      };
    }

    const best = detections.reduce((a, b) => (b.confidence > a.confidence ? b : a));
    const areaRatio = Math.min(1.0, detections.reduce((sum, d) => sum + d.bboxWidth * d.bboxHeight, 0));

    return {
      damageType: best.damageType,
      confidence: best.confidence,
      detections,
      damagedAreaRatio: round(areaRatio, 4),
      provider: this.name,
      modelName: this.modelName,
      modelVersion: this.modelVersion,
      processingMs: Math.round(performance.now() - started),
    };
  }

  async health(): Promise<boolean> {
    return true;
  }
}
